tests/integration/test_small_deployment_upgrades_edge_channel.py

```diff
@@ -268,8 +214,154 @@ async def test_upgrade_to_local(
             wait_for_exact_units={
                 APP_NAME: 3,
             },
+            timeout=3600,
             idle_period=IDLE_PERIOD,
         )
 
     # continuous writes checks
     await assert_continuous_writes_consistency(ops_test, c_writes, [app])
+
+
+##################################################################################
+#
+#  ROLLBACK test scenarios:
+#    Start with each version, moving to local and then rolling back mid-upgrade
+#
+##################################################################################
+
+
+ROLLBACK_PREFIX = "rollback_v"
+
+
+ROLLBACK_TEST_TYPE = [
+    (
+        pytest.param(
+            version,
+            id=f"{ROLLBACK_PREFIX}{version}",
+            marks=pytest.mark.group(f"{ROLLBACK_PREFIX}{version}"),
+        )
+    )
+    for version in VERSION_TO_REVISION.keys()
+]
+
+
+@pytest.mark.parametrize("version", ROLLBACK_TEST_TYPE)
```

Comment:
Why do we have separate groups? Why not all the upgrade / rollback flow in a single group?

Reply:
PS: there is no point at testing rollback of, e.g. 2.12.0 > 2.13.0 > 2.12.0, as this has been already tested when 2.13.0 was being released (i.e. it was the locally built charm)

Reply:
One thing to keep in mind, is that those integration tests will run for future versions too. So it's less about `2.12.0` or `2.13.0` but more about `N-1, ..., N-m`

Reply:
Yes, that is correct @Mehdi-Bendriss: as we add new versions, we will put them in the "version_to_revision" dict we have here.

My point was that, in a PR, we are only interested on how `upgrade from N-m to locally built` or `rollback from N-m to locally built` really play off.

There is an exception of what we call "the happy path", which tries to simulate the lifecycle of a deployment, starting on version `2.12.0` and doing all the right moves all the way to "locally built" revision.